### PR TITLE
Enable trimming for data protection

### PR DESCRIPTION
**Pull request by pranavkm**

### CODE REVIEW

**Comment by pranavkm** [comment]
DataProtection supports newing up types based on config strings e.g. https://github.com/dotnet/aspnetcore/blob/main/src/DataProtection/DataProtection/src/RegistryPolicyResolver.cs#L57-L60. This results in RequiresUnreferencedCode spreading virally in these APIs.

Additionally, there are two call sites where the DataProtection APIs are invoked from interface implementations. The trimmer requires that interfaces are annotated the same way implementations are. In this case we are unable to modify the interface (they are `IHostingStartup.RunAsync`, `IOptionsSetup.Configure`), so I used the xml file to suppress the warning from this repo.

**Comment by ShreyasJejurkar** [comment]
Why bot is adding community contributions to your PRs?

**Comment by mkArtakMSFT** [comment]
Thanks @pranavkm.
@adityamandaleeka can you please get this reviewed? Thanks!

**Comment by mkArtakMSFT** [comment]
@HaoK can you please review this? Thanks!

**Comment by HaoK** [comment]
@JamesNK are you handling all of the trimming PRs?

File: src/DataProtection/shared/src/TrimmerWarning.cs [hunk 1]
```diff
@@ -0,0 +1,9 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+namespace Microsoft.AspNetCore.DataProtection;
+
+internal static class TrimmerWarning
+{
+    public const string Message = "Data Protection may attempt to load types in a way that cannot be statically analyzed. Ensure all required types are preserved.";
```
Comment (JamesNK) [inline src/DataProtection/shared/src/TrimmerWarning.cs:8]: I think we need to improve this error message. "Ensure all required types are preserved." is what they need to do, but it doesn't give any indication about how they do it.

The simplest way to provide more info is to link to a doc. I took a look at https://docs.microsoft.com/en-us/dotnet/core/deploying/trimming/trim-self-contained and there isn't any content about what to do here.

Comment (pranavkm) [reply src/DataProtection/shared/src/TrimmerWarning.cs:8]: We could add trimming specific guidance to the [DataProtection docs](https://docs.microsoft.com/aspnet/core/security/data-protection/introduction?view=aspnetcore-6.0). In theory, the average user wouldn't run in to trimming specific issues with this library - it's only when they start configuring providers via xml configs or the registry that it runs the risk of being trimmed away.

Comment (JamesNK) [reply src/DataProtection/shared/src/TrimmerWarning.cs:8]: I agree.

This doesn't need to block this PR. Trimming docs issue: https://github.com/dotnet/aspnetcore/issues/41229

File: src/Shared/TrimmingAttributes.cs [hunk 2]
```diff
@@ -0,0 +1,120 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+#nullable enable
+
+namespace System.Diagnostics.CodeAnalysis;
+/// <summary>
```
Comment (JamesNK) [inline src/Shared/TrimmingAttributes.cs:7]:
```suggestion
namespace System.Diagnostics.CodeAnalysis;

/// <summary>
```

**Review by JamesNK — APPROVED** [review]

**Pull request merged**